executor/simple.go

```diff
@@ -820,85 +810,230 @@ func (e *SimpleExec) executeRollback(s *ast.RollbackStmt) error {
 		}
 	}
 	return nil
 }
 
-func (e *SimpleExec) authUsingCleartextPwd(authOpt *ast.AuthOption, authPlugin string) bool {
-	if authOpt == nil || !authOpt.ByAuthString {
-		return false
+func createUserFailedLoginJSON(info *passwordOrLockOptionsInfo) string {
+	if (info.failedLoginAttemptsChange || info.passwordLockTimeChange) && (info.failedLoginAttempts != 0 || info.passwordLockTime != 0) {
+		return fmt.Sprintf("{\"Password_locking\": {\"failed_login_attempts\": %d,\"password_lock_time_days\": %d}}",
+			info.failedLoginAttempts, info.passwordLockTime)
 	}
-	return mysql.IsAuthPluginClearText(authPlugin)
+	return ""
 }
 
-func (e *SimpleExec) isValidatePasswordEnabled() bool {
-	validatePwdEnable, err := e.ctx.GetSessionVars().GlobalVarsAccessor.GetGlobalSysVar(variable.ValidatePasswordEnable)
+func alterUserFailedLoginJSON(info *alterUserPasswordLocking, lockAccount string) string {
+	passwordLockingArray := []string{}
+	if lockAccount == "N" && (info.failedLoginAttempts != 0 || info.passwordLockTime != 0) {
+		passwordLockingArray = append(passwordLockingArray, fmt.Sprintf("\"auto_account_locked\": \"%s\"", lockAccount))
+		passwordLockingArray = append(passwordLockingArray, fmt.Sprintf("\"auto_locked_last_changed\": \"%s\"", time.Now().Format(time.UnixDate)))
+		passwordLockingArray = append(passwordLockingArray, fmt.Sprintf("\"failed_login_count\": %d", 0))
+	}
+	if info.failedLoginAttempts != 0 || info.passwordLockTime != 0 {
+		passwordLockingArray = append(passwordLockingArray, fmt.Sprintf("\"failed_login_attempts\": %d", info.failedLoginAttempts))
+		passwordLockingArray = append(passwordLockingArray, fmt.Sprintf("\"password_lock_time_days\": %d", info.passwordLockTime))
+	}
+	if len(passwordLockingArray) > 0 {
+		return fmt.Sprintf("\"Password_locking\": {%s}", strings.Join(passwordLockingArray, ","))
+	}
+	return ""
+}
+
+func readUserAttributes(ctx context.Context, sqlExecutor sqlexec.SQLExecutor, name string, host string, pLO *passwordOrLockOptionsInfo) (*alterUserPasswordLocking, error) {
+	alterUserInfo := &alterUserPasswordLocking{0, 0, false, false, false}
+	sql := new(strings.Builder)
+	sqlexec.MustFormatSQL(sql, `SELECT  JSON_UNQUOTE(JSON_EXTRACT(user_attributes, '$.Password_locking.failed_login_attempts')), JSON_UNQUOTE(JSON_EXTRACT(user_attributes, '$.Password_locking.password_lock_time_days')),
+	JSON_UNQUOTE(JSON_EXTRACT(user_attributes, '$.metadata'))          FROM %n.%n WHERE User=%? AND Host=%?;`, mysql.SystemDB, mysql.UserTable, name, strings.ToLower(host))
+	recordSet, err := sqlExecutor.ExecuteInternal(ctx, sql.String())
 	if err != nil {
-		return false
+		return nil, err
 	}
-	return variable.TiDBOptOn(validatePwdEnable)
+	rows, err := sqlexec.DrainRecordSet(ctx, recordSet, 3)
+	if err != nil {
+		return nil, err
+	}
+
+	if pLO.failedLoginAttemptsChange {
+		alterUserInfo.failedLoginAttempts = pLO.failedLoginAttempts
+	} else {
+		FailedLoginAttempts := rows[0].GetString(0)
```

Comment:
Local variables should start with lower-case, since upper- or lower-case can used to specify public or private in Golang.